### PR TITLE
creating CostAccountingBack

### PR DESCRIPTION
**Pull request by blokhinvalentin**

### CODE REVIEW

File: src/models/item.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+const { Schema } = mongoose;
+
+const itemSchema = new Schema({
+  where: String,
+  when: String,
+  cost: Number
+});
+
+module.exports = Item = mongoose.model('Items', itemSchema);
```
Comment (baklanovairina) [inline src/models/item.js:11]: Item, Items - too generic name

File: src/models/item.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+const mongoose = require('mongoose');
+
+const { Schema } = mongoose;
+
+const itemSchema = new Schema({
+  where: String,
```
Comment (baklanovairina) [inline src/models/item.js:6]: maybe "place"?

File: src/models/item.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  when: String,
```
Comment (baklanovairina) [inline src/models/item.js:7]: time or date

and type: Date

File: app.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+app.use("/", apiRoutes);
+
+const launch = () => {
+  try {
+    mongoose.connect(process.env.URL, { useNewUrlParser: true, useUnifiedTopology: true });
+    app.listen(8000, () => {
+      console.log('connected to port 8000');
+    });
+  }
+  catch(error) {
```
Comment (baklanovairina) [inline app.js:21]: add space after catch

File: package.json [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "start": "nodemon node app.js"
+  },
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "cors": "^2.8.5",
+    "express": "^4.17.1",
+    "nodemon": "^2.0.7"
```
Comment (baklanovairina) [inline package.json:15]: where dotenv and mongoose?

File: src/models/item.js (hunk 2 above)
Comment (baklanovairina) [reply src/models/item.js:6]: if this is required field, then can add 
`required: true`

File: src/models/item.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+  cost: Number
```
Comment (baklanovairina) [inline src/models/item.js:8]: if this is required field, then can add 
`required: true`

File: src/models/item.js (hunk 3 above)
Comment (baklanovairina) [reply src/models/item.js:7]: can add default value:

`default: Date.now`

File: src/modules/controllers/item-controller.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
+const Item = require('../../models/item');
+
+const getItems = (req, res) => {
+  try {
+    Item.find().then(result => {
+      res.status(200).send(result);
+    });
+  }
+  catch(error) {
```
Comment (baklanovairina) [inline src/modules/controllers/item-controller.js:9]: add space

File: src/modules/controllers/item-controller.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
+const Item = require('../../models/item');
+
+const getItems = (req, res) => {
```
Comment (baklanovairina) [inline src/modules/controllers/item-controller.js:3]: about the name wrote above

File: src/modules/controllers/item-controller.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+  catch(error) {
+    res.status(400).send('can get all items');
+  }
+}
+
+const addItem = (req, res) => {
+  try {
+    const { where, when, cost } = req.body;
+
+    if (where === '' || cost === '') {
```
Comment (baklanovairina) [inline src/modules/controllers/item-controller.js:18]: add data type check
"where" should be "string" 
"cost" should be "number"

File: src/modules/controllers/item-controller.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+      throw new Error('something is empty');
+    }
+
+    const item = new Item({ where, when, cost });
```
Comment (baklanovairina) [inline src/modules/controllers/item-controller.js:22]: why "when"? It is created by default.

File: src/modules/controllers/item-controller.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    item.save().then(() => {
+      getItems(req, res);
+    });
+  }
+  catch(error) {
```
Comment (baklanovairina) [inline src/modules/controllers/item-controller.js:27]: space

File: src/modules/controllers/item-controller.js [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    res.status(400).send('can add item');
+  }
+}
+
+const editItem = (req, res) => {
+  try {
+    const { _id, where, when, cost } = req.body;
+    if (!req.body.hasOwnProperty('_id') || where === '' || when === '' || cost === '') {
```
Comment (baklanovairina) [inline src/modules/controllers/item-controller.js:35]: you also need to check the type (where, when, cost)

File: src/modules/controllers/item-controller.js [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+      throw new Error('unable to update text');
+    }
+
+    Item.updateOne(
+      { _id }, 
+      { $set: { where: where, when: when, cost: cost } }
```
Comment (baklanovairina) [inline src/modules/controllers/item-controller.js:41]: when the key and value are named the same, the abbreviation can be used:
{ $set: { where, when, cost } }
and you forgot _id

File: src/modules/controllers/item-controller.js [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+      ).then(() => {
+      getItems(req, res);
+    });
+  }
+  catch(error) {
```
Comment (baklanovairina) [inline src/modules/controllers/item-controller.js:46]: space

File: src/modules/controllers/item-controller.js [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+      throw new Error('id is unreachable to read');
+    }
+
+    const _id = req.body._id;
+    
+    Item.deleteOne({ _id }).then(result => {
+      res.status(200).send(result);
+    })
+  }
+  catch(error) {
```
Comment (baklanovairina) [inline src/modules/controllers/item-controller.js:63]: space

File: app.js (hunk 4 above)
Comment (baklanovairina) [inline app.js:21]: `} catch (error) {`

File: app.js [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+    console.log(error);
```
Comment (baklanovairina) [inline app.js:22]: console.error

File: src/models/item.js [hunk 17]
```diff
@@ -0,0 +1,20 @@
+const mongoose = require('mongoose');
+
+const { Schema } = mongoose;
+
+const spendingsSchema = new Schema({
+  where: {
```
Comment (baklanovairina) [inline src/models/item.js:6]: use nouns for names

File: src/models/item.js [hunk 18]
```diff
@@ -0,0 +1,20 @@
+const mongoose = require('mongoose');
+
+const { Schema } = mongoose;
+
+const spendingsSchema = new Schema({
+  where: { 
+    type: String,
+    required: true
+  },
+  when: {
```
Comment (baklanovairina) [inline src/models/item.js:10]: use nouns for names

File: src/models/item.js [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+    type: Date  ,
```
Comment (baklanovairina) [inline src/models/item.js:11]: delete space

File: src/models/item.js [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+    default: Date.now()
+  },
+  cost: {
+    type: Number,
+    required: true
+  }
+});
+
+module.exports = Spending = mongoose.model('MySpendings', spendingsSchema);
```
Comment (baklanovairina) [inline src/models/item.js:20]: without "My"

File: src/modules/controllers/item-controller.js [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+    res.status(400).send('can get all items');
+  }
+}
+
+const addItem = (req, res) => {
+  try {
+    const { where, cost } = req.body;
+
+    if ((where === '') || (typeof where !== 'string')
+    || (cost === '') || (typeof cost !== 'number')) {
```
Comment (baklanovairina) [inline src/modules/controllers/item-controller.js:18]:
```
if ((where === '') 
  || (typeof where !== 'string')
  || (cost === '') 
  || (typeof cost !== 'number')
) {
```

File: src/modules/controllers/item-controller.js [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+      throw new Error('something is empty');
```
Comment (baklanovairina) [inline src/modules/controllers/item-controller.js:19]: here there is a check for the data type, not only for an empty field

File: src/modules/controllers/item-controller.js [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    const item = new Spending({ where, cost });
+    item.save().then(() => {
```
Comment (baklanovairina) [inline src/modules/controllers/item-controller.js:23]:
```
item.save().then(result => {
   res.status(200).send(result);
});
```

File: src/modules/controllers/item-controller.js [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+const editItem = (req, res) => {
+  try {
+    const { _id, where, when, cost } = req.body;
+
+    if (!req.body.hasOwnProperty('_id') 
+      || where === '' 
+      || when === '' 
+      || cost === ''
+      || typeof where !== 'string'
+      || typeof cost !== 'string'
```
Comment (baklanovairina) [inline src/modules/controllers/item-controller.js:40]: You often repeat validation. Move it to a separate function and place it in a separate "validation.js" file. Move the file to the src/helpers folder

Comment (baklanovairina) [reply src/modules/controllers/item-controller.js:40]: And here this function is imported and called

File: src/modules/controllers/item-controller.js [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+      throw new Error('unable to update text');
+    }
+
+    Spending.findOneAndUpdate(
+      { _id }, 
+      { $set: { where, when, cost } },
+      { new: true }
+      ).then(result => {
+        res.status(200).send(result);
+    });
```
Comment (baklanovairina) [inline src/modules/controllers/item-controller.js:51]: tab

File: src/modules/controllers/item-controller.js [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+      res.status(200).send(item);
+    });
+  } catch (error) {
+    res.status(400).send('can add item');
+  }
+}
+
+const editItem = (req, res) => {
+  try {
+    const { _id, where, when, cost } = req.body;
```
Comment (baklanovairina) [inline src/modules/controllers/item-controller.js:33]: _id send through "params"

File: src/modules/controllers/item-controller.js [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+  }
+}
+
+const deleteItem = (req, res) => {
+  try { 
+    if (!req.body.hasOwnProperty('_id')) {
+      throw new Error('id is unreachable to read');
+    }
+
+    const _id = req.body._id;
```
Comment (baklanovairina) [inline src/modules/controllers/item-controller.js:63]: use params

File: src/modules/controllers/item-controller.js (hunk 22 above)
Comment (blokhinvalentin) [reply src/modules/controllers/item-controller.js:19]: why do I need its check if i set date type and value by default

File: src/helpers/validation.js [hunk 28]
```diff
@@ -0,0 +1,8 @@
+const stringValidation = (string) => typeof string === 'string' || string !== '';
+
+const numberValidation = (number) => typeof number === 'number' || number !== null;
```
Comment (baklanovairina) [inline src/helpers/validation.js:3]: why null?

File: src/modules/controllers/spendings-controller.js [hunk 29]
```diff
@@ -0,0 +1,80 @@
...
+
+    Spending.findOneAndUpdate(
+      { _id }, 
+      { $set: { place, time, cost } },
+      { new: true }
+    ).then(result => {
+      res.status(200).send(result);
+    });
+  } catch (error) {
+    res.status(400).send('cant edit item');
```
Comment (baklanovairina) [inline src/modules/controllers/spendings-controller.js:55]: item?

File: src/modules/controllers/spendings-controller.js [hunk 30]
```diff
@@ -0,0 +1,80 @@
...
+      throw new Error('id is unreachable to read');
+    }
+
+    const _id = req.params._id;
+    
+    Spending.findOneAndDelete({ _id }).then(result => {
+      res.status(200).send(result);
+    })
+  } catch (error) {
+    res.status(400).send('cant delete item');
```
Comment (baklanovairina) [inline src/modules/controllers/spendings-controller.js:71]: item?

**Pull request merged**